README
WDB2R
=====

R-scripts for extracting and converting data from the wdb system
Before running this scripts install R, and then install packages DBI, RPostgreSQL and udunits

Installation:
-------------
sudo apt-get install libudunits2-dev 

in R:
> install.packages("DBI")
> install.packages("RPostgreSQL")
> install.packages("udunits2")

Use
----

> source("readVerifWdb.R")

The function
readVerifWdb<-function(wmo_no,period,model,prm,prg,testMemory=false,testLatency=false)
extracts data for the given station number, time periods, parameters, models and prognosis lengths.

Output is an R.dataframe with a format similar to this
  wmo_no       time prog   DD.H8 FF.H8   TT.H8     P.H8  DD.UM4 FF.UM4   TT.UM4  P.UM4
1   10380 2011022201    1 173.824 3.358 260.137 1029.920 174.824 3.356 260.100 1029.800
2   10380 2011022202    2 178.778 3.506 259.849 1029.620 178.777 3.505 259.169 1029.610
3   10380 2011022203    3 181.600 3.231 259.648 1029.420 182.710 3.230 259.523 1029.420

with all values for one place/time/prog  on one row

call finish() to disconnect

Examples of use
---------------
> mydf<-readVerifWdb(c(1317),c(20120128,20120130),c("H4","H12"),c("TT","P"),NULL)
> mydf<-readVerifWdb(wmo_no=c(1492),period=c(20090801,20090805),model=c("EC","UM4","H12"),prm=c("FF","DD"),prg=c(3,6,9))
> mydf<-readVerifWdb(c(1003),c(20120131,20120203),c("H8","EC"),c("TD","RR"), c(18,36))


> source("testWdb.R")
# find all models(shortmodelname) for station 1493, 1/3-2010, parameter wind speed(FF):
> findAvailableModels(1493,20100301,"FF")

# find all parameters(miopdb_par) for station 1493, 1/3-2010, model Hirlam 8
> findAvailableParameters(1493,20100301,"H8")

# find all stations(wmo_no) for  1/3-2010, parameter wind speed(FF), model Hirlam 8
> findAvailablePlaces(20100301,"FF","H8")

# find all referencetimes for station 1493, parameter wind speed(FF), model Hirlam 8
>  findAvailableReferencetimes(1493,"FF","H8")
#
# find all proglenghts for station 1493,1/3-2010, parameter wind speed(FF), model Hirlam 8
> findAvailableProglengths(1493,20100301,"FF","H8")


# vi: se ft=markdown:


debian/changelog
wdb2r-1.0 (1.0) precise; urgency=low

  * Initial release.

 -- Kim Tore Jensen <kimtj@met.no>  Wed, 08 Jan 2014 11:14:34 +0100


testWdb.R
source("readVerifWdb.R")

# Examples of use
#
# find all models(shortmodelname) for station 1493, 1/3-2010, parameter wind speed(FF):
# findAvailableModels(1493,20100301,"FF")
#
# find all parameters(miopdb_par) for station 1493, 1/3-2010, model Hirlam 8
# findAvailableParameters(1493,20100301,"H8")
#
# find all stations(wmo_no) for  1/3-2010, parameter wind speed(FF), model Hirlam 8
# findAvailablePlaces(20100301,"FF","H8")
#
# find all referencetimes for station 1493, parameter wind speed(FF), model Hirlam 8
# findAvailableReferencetimes(1493,"FF","H8")
#
# find all proglenghts for station 1493,1/3-2010, parameter wind speed(FF), model Hirlam 8
# findAvailableProglengths(1493,20100301,"FF","H8")

findAvailableModels<-function(wmo_no=NULL,date=NULL,prm=NULL){
# returns vector of available models
  startup()
  query <- "SELECT dataprovidername FROM wci.browse( NULL,WMO_NO ,TIME,NULL, PARAMETER,NULL, NULL,NULL::wci.browsedataprovider )"
  query <- sub("WMO_NO",wmostring(wmo_no) ,query)
  query <- sub("TIME",timestring(date),query)
  query <- sub("PARAMETER",parameterstring(prm),query)
  print(query)
  rs <- dbSendQuery(con, query)
  results<<-fetch(rs,n=-1)
  if (nrow(results)==0){
    return()
  }
  dbClearResult(rs)
  vmodels<<-vector(length=0)
  if (nrow(results)==0)
    return(vmodels)
  for (i in 1:nrow(results)){
    dp <- results$dataprovidername[i]
    dpdef <- dataproviderDefinitions[dataproviderDefinitions$dataprovider==dp,]
    if (nrow(dpdef)==1){
      mod<-dpdef$shortmodelname
      vmodels<<-c(vmodels,as.character(mod))
    } else {
      vmodels<<-c(vmodels,as.character(dp))
    }
  }
  return(vmodels)
 }


findAvailableParameters<-function(wmo_no=NULL,date=NULL,model=NULL){
  startup()
 # returns vector of available parameters
  query <- "SELECT valueparametername FROM wci.browse(DATAPROVIDER,WMO_NO ,TIME,NULL, NULL,NULL, NULL,NULL::wci.browsevalueparameter )"
  query <- sub("WMO_NO",wmostring(wmo_no) ,query)
  query <- sub("TIME",timestring(date),query)
  query <- sub("DATAPROVIDER",dataproviderstring(model),query)
  print(query)
  rs <- dbSendQuery(con, query)
  results<<-fetch(rs,n=-1)
  dbClearResult(rs)
  vparams<<-vector(length=0)
  if (nrow(results)==0)
    return(vparams)
  for (i in 1:nrow(results)){
    vp <- results$valueparametername[i]
    vpdef <- parameterDefinitions[parameterDefinitions$valueparametername==vp,]
    if (nrow(vpdef)>0){
      for (j in 1:nrow(vpdef)){
        par<-vpdef[j,]$miopdb_par
        vparams <<- c(vparams,as.character(par))
      }
    } else {
      vparams<<-c(vparams,as.character(vp))
    }
  }
  return(vparams)
}


findAvailablePlaces<-function(date=NULL,prm=NULL,model=NULL){
 # returns vector of available places
  startup()
  query <- "SELECT placename FROM wci.browse( DATAPROVIDER,NULL ,TIME,NULL, PARAMETER,NULL, NULL,NULL::wci.browseplace ) order by placename"  
  query <- sub("TIME",timestring(date),query)
  query <- sub("DATAPROVIDER",dataproviderstring(model),query)
  query <- sub("PARAMETER",parameterstring(prm),query)  
  print(query)
  rs <- dbSendQuery(con, query)
  results<-fetch(rs,n=-1)
  dbClearResult(rs)
  vplaces<-vector(length=nrow(results))
  for (i in 1:nrow(results)){
    vplaces[i]<-results$placename[i]
  }
  return(vplaces)
 }


findAvailableReferencetimes<-function(wmo_no=NULL,prm=NULL,model=NULL){
 # returns vector of available reftimes
  query <- "SELECT distinct(to_char(referencetime,'YYYYMMDDHH24')) as time FROM wci.browse( DATAPROVIDER,WMO_NO ,NULL,NULL, PARAMETER,NULL, NULL,NULL::wci.browsereferencetime) order by time"        
  query <- sub("DATAPROVIDER",dataproviderstring(model),query)
  query <- sub("PARAMETER",parameterstring(prm),query)
  query <- sub("WMO_NO",wmostring(wmo_no) ,query)
  print(query)
  rs <- dbSendQuery(con, query)
  results<-fetch(rs,n=-1)
  dbClearResult(rs)
  vreftimes<-vector(length=nrow(results))
  for (i in 1:nrow(results)){
    vreftimes[i]<-results$time[i]
  }
  vreftimes<-sort(vreftimes)
  return(vreftimes)

}


findAvailableProglengths<-function(wmo_no=NULL,date=NULL,prm=NULL,model=NULL){
 # returns vector of available progtimes
  query<-"SELECT distinct(wci.prognosishour(referencetime, validtimefrom)) as prog FROM wci.read(DATAPROVIDER,WMO_NO,TIME,NULL, PARAMETER,NULL, NULL,NULL::wci.returnfloat) order by prog"
  query <- sub("WMO_NO",wmostring(wmo_no) ,query)
  query <- sub("TIME",timestring(date),query)
  query <- sub("DATAPROVIDER",dataproviderstring(model),query)
  query <- sub("PARAMETER",parameterstring(prm),query)
  print(query)
  rs <- dbSendQuery(con, query)
  results<-fetch(rs,n=-1)
  dbClearResult(rs)
  vprogs<-vector(length=nrow(results))
  for (i in 1:nrow(results)){
    vprogs[i]<-results$prog[i]
  }

  return(sort(vprogs))
 }



# utility functions
wmostring <- function(wmo_no){
  if (is.null(wmo_no)){
    wmostring <- "NULL"    
  } else{
    wmostring <- paste("'",wmo_no,"'",sep="")
  }
  return(wmostring)
}

timestring <- function(date){
  if (is.null(date)){
    timestring <- "NULL"
  } else{
    time <- getFormattedTime(date)
    timestring <- paste("'",time,"'",sep="")
  }
  return(timestring)
}

parameterstring <- function(prm){
  if (is.null(prm)){
    parameterstring <- "NULL"    
  } else{
    pdef <- parameterDefinitions[parameterDefinitions$miopdb_par==prm,]
    parameter <- as.character(pdef$valueparametername)
    parameterstring <- paste("'{",parameter,"}'",sep="")
  }
  return(parameterstring)
}

dataproviderstring <- function(model){  
  if (is.null(model)){
    dataproviderstring <- "NULL";
  } else {  
    dpdef <- dataproviderDefinitions[dataproviderDefinitions$shortmodelname==model,]
    dataprovider <- dpdef$dataprovider
    dataproviderstring <- paste("ARRAY['",dataprovider,"']",sep="")
  }
  return(dataproviderstring)
}


readVerifWdb.R
# before running this script install packages from CRAN
# in R:
# > install.packages("DBI")
# > install.packages("RPostgreSQL")
# > install.packages("udunits2")
#

# Example of use:
# connect to database:
# > startup(user, host, database)
# run queries:
# > mydf<-readVerifWdb(c(1003),c(20120131,20120203),c("EC"),c("TT","P"), c(18,36))
# > mydf<-readVerifWdb(c(1493),c(20100131,20100203),c("EC","UM4"),c("FF","DD"), NULL)
# > mydf<-readVerifWdb(c(1317),c(1990128,19990130),c("H50","H10"),c("DD","FF"),prg=seq(3,21,3))
# ... more calls to readVerifWdb ...
# close connecion when finished
# > finish()



# define parameters
DD<-c("DD","degrees","wind from direction","air pressure","height above ground","10","radians")
FF<-c("FF","m/s","wind speed","air pressure","height above ground","10","m/s")
TT<-c("TT","celsius","air temperature","air pressure","height above ground","2","kelvin")
TT_K<-c("TT_K","celsius","kalman air temperature","air pressure","height above ground","2","kelvin")
TT.K<-c("TT.K","celsius","kalman air temperature","air pressure","height above ground","2","kelvin")
TD<-c("TD","celsius","dew point temperature","air pressure","height above ground","2","kelvin")
RR<-c("RR","mm","lwe thickness of precipitation amount","air pressure","height above ground","0","m")
P<-c("P","hPa","air pressure at sea level",NA,"height above ground","0","Pa")
N<-c("N","%","cloud area fraction",NA,"air pressure","100000","%")
CH<-c("CH","%","cloud area fraction in atmosphere layer",NA,"air pressure","30000","%")
FG<-c("FG","%","cloud area fraction in atmosphere layer",NA,"air pressure","100000","%")
CL<-c("CL","%","cloud area fraction in atmosphere layer",NA,"air pressure","85000","%")
CM<-c("CM","%","cloud area fraction in atmosphere layer",NA,"air pressure","50000","%")
UU<-c("UU","%","relative humidity","air pressure",NA,NA,"%")
Z<-c("Z",NA,"atmosphere sigma coordinate","air pressure",NA,NA,NA)
parameterDefinitions<-data.frame(rbind(DD,FF,TT,TT_K,TT.K,TD,RR,P,N,CH,FG,CL,CM,UU,Z))
colnames(parameterDefinitions)<-c("miopdb_par","miopdb_unit","valueparametername","levelparametername","groundlevelparametername","groundlevel","unit")

NIVA<-c("NIVA","hPa","air pressure","Pa")
levelparameterDefinitions <-data.frame(rbind(NIVA))
colnames(levelparameterDefinitions)<-c("miopdb_par","miopdb_unit","levelparametername","unit")


# definitions of dataproviders
hirlam50<-c("hirlam50km_interpolated_v0","H50")
hirlam10<-c("hirlam10km_interpolated_v0","H10")
hirlam20<-c("hirlam20km_interpolated_v0","H20")
hirlam4<-c("hirlam4km_interpolated_v0","H4")
hirlam8<-c("hirlam8km_interpolated_v0","H8")
hirlam12<-c("hirlam12km_interpolated_v0","H12")
EC<-c("ecmwf15_interpolated_v0","EC1_LOW")
ECMWF<-c("ecmwf_interpolated_v0","EC")
UK<-c("uk_global_interpolated_v0","UK")
UM4KM1<-c("unified_model4km_interpolated_v0","UM4")
dataproviderDefinitions<-data.frame(rbind(hirlam50,hirlam10,hirlam20,hirlam4,hirlam8,hirlam12, EC,ECMWF,UK,UM4KM1))
colnames(dataproviderDefinitions)<-c("dataprovider","shortmodelname")

started <<-FALSE 

startup<-function(user, host, dbname){
  if (started)
    return(TRUE)
  if (!require(udunits2)) {
    cat("The udunits2 package is not available.\n")
    return(FALSE)
  }   
  if (!require(DBI)) {
    cat("The DBI package is not available.\n")
    return(FALSE)
  }       
  if (!require(RPostgreSQL)) {
    cat("The RpostgreSQL package is not available.\n")
    return(FALSE)
  }
 #wdb stuff  
  namespace <- "88,42,88"
 #startup; load driver, connect to wdb etc
  drv<<-dbDriver("PostgreSQL")
  con<<-dbConnect(drv,  dbname=dbname,  user=user, host=host)
  sendquery <- "select wci.begin('wdb',88,42,88)"
  rs<-suppressWarnings(dbSendQuery(con,sendquery))
  dbClearResult(rs)
  started<<-TRUE
  return(TRUE)
}


finish<-function(){
 # finish; release wci, disconnect, unload driver
  rs<-suppressWarnings(dbSendQuery(con,"select wci.end()"))
  dbClearResult(rs)
  dbDisconnect(con);
  dbUnloadDriver(drv)
}


readVerifWdb<-function(wmo_no,period,model,prm,prg,lev=NULL,init.time=0,useReftime=FALSE){

  if (!started){
    cat("Please call startup(user, host, database) before using this function.\n")
    return(FALSE)
  }
  
  if (length(wmo_no)!=1){
    cat("Length of wmo_no!=1\n")
    return()
  }

  queryPart1<-"select * from (select placename as WMO_NO, to_char(validtimefrom,'YYYYMMDDHH24') as TIME, wci.prognosishour(referencetime, validtimefrom)  AS PROG"
  if (!is.null(lev))
    queryPart1 <- paste(queryPart1,",levelfrom as LEV")

  
  #create empty data frame to hold all models
  allmodels<-data.frame()
# loop over models
  for (mod in model){
    for (par  in prm){

    # each model/parameter combo results in a dataframe with rows like this
    # WMO_NO, TIME,PROG, TT.EC
    # which are then merged together to something like
    # WMO_NO, TIME,PROG, TT.EC TT.UM FF.EC FF.UM        
      modelString<-getDataProviderString(mod)
      locationString <-paste("'",wmo_no,"'",sep="")
      if (useReftime){
        reftimeString<-getInsideTimeString(period)
        validtimeString<-"NULL"
      } else{
        reftimeString<-getInsideTimeString(period,subtracthours=300)
        validtimeString<-getInsideTimeString(period)
      }
      parameterString<-getParameterString(par)
      levelString<-getLevelString(par,lev)
      versionString<-"NULL"
      returnString<-"NULL::wci.returnfloat)"
      init.timestring <- sprintf("%02i", init.time)
      terminString <- paste("where to_char(referencetime,'HH24') in ('",init.timestring,"')",sep="")
      whereString<-")AS wciquery "
      if(!is.null(prg)){
        progString<-paste("where (PROG in(",paste(prg,collapse=","),") )",sep="")      
      } else {
        progString <- ""
      }
      parmod <- paste(par,mod,sep=".")
      parmodel<-paste("\"", parmod  ,"\"",sep="")
      queryPart2<- paste(",value as",parmodel, "from wci.read(")
      queryPart3<-paste(modelString,locationString,reftimeString,validtimeString,parameterString,levelString,versionString,returnString,sep=",")
      queryPart4<-paste(terminString,whereString,progString,"order by TIME")
      
      query<-paste(queryPart1,queryPart2,queryPart3,queryPart4)

      cat(query,"\n") 
      rs <- dbSendQuery(con, query)
      results<-fetch(rs,n=-1)
 
      cat("Number of rows in results", nrow(results),"\n")
      cat("Size of results", object.size(results),"bytes \n")

      dbClearResult(rs)
     
      if (nrow(results)!=0){
        values<-as.double(unlist(results[parmod]))
        #convert values to other units
        values <- convertValues(values,par)
        results[parmod] <- values
        levs <- results$lev
        if (length(levs)!=0){
          levs <- convertLevels(levs,par)
          results$lev <- levs
        }
      }
          
      if (nrow(allmodels)==0){      
        allmodels<-results
      }
      else{
        if (nrow(results)!=0){      
          allmodels<-merge(allmodels,results)
        }
        
      }
      
    }
  }
  
    
  return(allmodels)  
}



readVerifWdbMultipleStations<-function(wmo_no,period,model,prm,prg,lev=NULL,init.time=0,useReftime=FALSE,testMemory=FALSE,testLatency=FALSE){

  if (!started){
    cat("Please call startup(user, host, database) before using this function.\n")
    return(FALSE)
  }
  
  queryPart1<-"select * from (select placename as WMO_NO, to_char(validtimefrom,'YYYYMMDDHH24') as TIME, wci.prognosishour(referencetime, validtimefrom)  AS PROG"
  if (!is.null(lev))
    queryPart1 <- paste(queryPart1,",levelfrom as LEV")

  maxmemory<<-0
  #create empty data frame to hold all models
  allmodels<-data.frame()
# loop over models
  for (mod in model){
    for (par  in prm){

    # each model/parameter combo results in a dataframe with rows like this
    # WMO_NO, TIME,PROG, TT.EC
    # which are then merged together to something like
    # WMO_NO, TIME,PROG, TT.EC TT.UM FF.EC FF.UM      
   
      modelString<-getDataProviderString(mod)
      locationString <-"NULL"
      if (useReftime){
        reftimeString<-getInsideTimeString(period)
        validtimeString<-"NULL"
      } else{
        reftimeString<-getInsideTimeString(period,subtracthours=300)
        validtimeString<-getInsideTimeString(period)
      }
      parameterString<-getParameterString(par)
      levelString<-getLevelString(par,lev)
      versionString<-"NULL"
      returnString<-"NULL::wci.returnfloat)"
      init.timestring <- sprintf("%02i", init.time)
      terminString <- paste("where to_char(referencetime,'HH24') in ('",init.timestring,"')",sep="")
      whereString<-")AS wciquery where("
      stationString<-paste("WMO_NO in('",paste(wmo_no,collapse="','"),"')",sep="")      
      progString<-paste("AND PROG in(",paste(prg,collapse=","),") )",sep="")      
      
      parmod <- paste(par,mod,sep=".")
      parmodel<-paste("\"", parmod  ,"\"",sep="")
      queryPart2<- paste(",value as",parmodel, "from wci.read(")
      queryPart3<-paste(modelString,locationString,reftimeString,validtimeString,parameterString,levelString,versionString,returnString,sep=",")
      queryPart4<-paste(terminString,whereString,stationString,progString,"order by TIME")

      query<-paste(queryPart1,queryPart2,queryPart3,queryPart4)

      
      
      cat(query,"\n") 
      if (testLatency){
        ts<-system.time(rs <- dbSendQuery(con, query))
        tr<-system.time(results<-fetch(rs,n=-1))
        cat("Time for rs <- dbSendQuery(con, query)\n")
        print(ts)
        cat("Time for results<-fetch(rs,n=-1)\n")
        print(tr)
      }
      else{
        rs <- dbSendQuery(con, query)
        results<-fetch(rs,n=-1)
      }

      cat("Number of rows in results", nrow(results),"\n")
      cat("Size of results", object.size(results),"bytes \n")

      dbClearResult(rs)
     
#      finish()
                                       
      if (nrow(results)!=0){
        values<-as.double(unlist(results[parmod]))
        #convert values to other units
        values <- convertValues(values,par)
        results[parmod] <- values
        levs <- results$lev
        if (length(levs)!=0){
          levs <- convertLevels(levs,par)
          results$lev <- levs
        }
      }
          
      if (nrow(allmodels)==0){      
        allmodels<-results
      }
      else{
        if (testLatency){
          tm<-system.time(allmodels<-merge(allmodels,results))
          cat("Time for allmodels<-merge(allmodels,results) \n")
          print(tm)
        }
        else
          if (nrow(results)!=0)      
          allmodels<-merge(allmodels,results)
      }


      #monitor memory usage
      if (testMemory){
        myobj<-sort( sapply(ls(),function(x){object.size(get(x))}), decreasing=TRUE)
        totalmem<-sum(myobj)
        cat("Total memory", totalmem,"bytes \n")
        myobj<-head(myobj,5)
        cat("5 largest objects:\n")
        print(myobj)
        if (totalmem > maxmemory)
          maxmemory<<-totalmem
      }
      
      
    }
 
  }

  
  return(allmodels)  
}


readEnsembleWdb<-function(wmo_no,period,prm,model,prg,lev=NULL,members=NULL){
  # example call
  # readEnsembleWdb(wmo_no=c(10380,98790),period=c(20120310,20120312),model=c("EPS"),parameters=c("TT"),prg=c(12,18))->mydf
  # readEnsembleWdb(wmo_no=c(18700),period=c(20120310,20120312),model=c("EPS"),parameters=c("TCC"),prg=c(12),members=c(0:10))-> mydf
  
  # selection of ref and valid time, same for all queries
  reftimeString<-gsub("'","''",getInsideTimeString(period))
  validtimeString<-"NULL"
  locationString <-"NULL"
  
  # selection of stations and progs, same for all queries
  whereString<-" AS wciquery where "
  orderString <-  " order by rowid,dataversion'"
  stationString<-paste("wmo_no in(''",paste(wmo_no,collapse="'',''"),"'')",sep="")
  progString<-paste("AND prog in(",paste(prg,collapse=","),")",sep="")
  queryPart4<-paste(whereString,stationString,progString,orderString)

  if (is.null(members))
    # select all versions
    versionString="NULL"
  else
    versionString<-paste("ARRAY[", paste(members,collapse=","),"]",collapse="")
 
  
  allmodels<-data.frame()
  
  for (mod in model){
    for (par  in prm){
    # each model/parameter combo results in a dataframe with rows like this
    # WMO_NO, TIME,PROG, TT.EC.1 TT.EC.2 TT.EC.3 ........       TT.EC.50
    # which are then merged together to something like
    # WMO_NO, TIME,PROG, TT.EC.1 TT.EC.T2....  TCC.EC.1 TCC.EC.2 TCC.EC3...


      # create source_sql string from wmo_no, period, models, prm, prg
      # query which gives all the data + an extra row id column
      # the row id is for classification
      queryPart1 <-"'select * from (select (placename||'',''||to_char(validtimefrom,''YYYYMMDDHH24'')||'',''||wci.prognosishour(referencetime, validtimefrom)||'',''||levelfrom) as rowid , placename as wmo_no, to_char(validtimefrom,''YYYYMMDDHH24'') as time, wci.prognosishour(referencetime, validtimefrom) as prog,
  dataversion,  to_char(value,''9999D99'') "
      queryPart2<- "from wci.read("
      modelString<-gsub("'","''",getDataProviderString(mod))
      parameterString<-gsub("'","''",getParameterString(par))
      levelString <- gsub("'","''",getLevelString(par,lev))
      returnString<-"NULL::wci.returnfloat)"
      queryPart3<-paste(modelString,locationString,reftimeString,validtimeString,parameterString,levelString,versionString,returnString,sep=",")
      source_sql<-paste(queryPart1,queryPart2,queryPart3,")",queryPart4)
      
      # query which gives the version
      queryCategoryPart1 <- "' SELECT distinct(dataversion)";
      category_sql <-paste(queryCategoryPart1,queryPart2,queryPart3,"order by dataversion'")


      if (is.null(members)){
        # find number of dataversions=ensemble members 
        queryCountPart1 <- "SELECT count(distinct(dataversion))"
        count_sql <-gsub("''","'",paste(queryCountPart1,queryPart2,queryPart3))
        cat("\n",count_sql,"\n")
        rs <- dbSendQuery(con, count_sql )
        results<-fetch(rs,n=-1)
        nversion<- results[1,1]-1
        members<-c(0:nversion)
      }
        
      # columns to sort into, depends on how many ensembleversions we have    
      parmod <- paste(par,mod,sep=".")
      parmodel<- paste("\"",parmod,".",sep="")
      floatparmodel<-paste("\" float,", parmodel)
      colstring <- paste(members,collapse=floatparmodel)
      colstring <- paste(parmodel,colstring,"\" float",sep="")
      asct_sql <- paste("as ct(rowid text,wmo_no text, time text, prog int,",colstring,  ")")

      # the final crosstab query
      query<-paste("select * from crosstab(",source_sql, ",", category_sql, ")", asct_sql,sep="")

      cat("\n",query,"\n") 
      rs <- dbSendQuery(con, query)
      results<-fetch(rs,n=-1)
      results<-results[,-1]



      if (nrow(results)!=0){
        for (i in 1:ncol(results)){
          #check if this is a parameter column
          if (length(grep(parmod,names(results)[i],TRUE,TRUE))>0) {
            values<-as.double(unlist(results[,i]))
            #convert values to other units
            values <- convertValues(values,par)
            results[,i] <- values
          }
        }
      }
      
      if (nrow(allmodels)==0){      
        allmodels<-results
      }
      else{
        allmodels<-merge(allmodels,results)
      }
      
      
    }
  }


  return(allmodels)
 #remove first column which is just to sort by
 #merge models, parameters
}




getInsideTimeString<-function(period,subtracthours=0){
  if (is.null(period)){
    #use last week
    today<-Sys.Date()
    lastweek<-today-6
    period<-c(format(lastweek,"%Y%m%d"),format(today,"%Y%m%d"))
  }
  if (length(period)==1) period=c(period,period)	
  # now length of period is at least two (use only the first two entries)
  fstarttime<-getFormattedTime(period=period[1],subtracthours=subtracthours)
  fendtime<-getFormattedTime(period=period[2],addDay=TRUE)
  rtstring<-paste("'inside ",fstarttime," TO ",fendtime,"'",sep="")
  return(rtstring)
}


getFormattedTime<-function(period,addDay=FALSE,subtracthours=0){
  # get formatted time ie 2012-01-28 00:00:00  from period ie 20120128
  # if addDay=true, add one day
  timestring<-paste(period,00)
  time<-strptime(timestring,"%Y%m%d%H")
  if (addDay)time<-time+23*3600
  if (subtracthours!=0) time <- time-subtracthours*3600
  ftime<-format(time,"%Y-%m-%d %H:%M:%S")
  return(ftime)
}

getDataProviderString<-function(model){
  dataprovider<-dataproviderDefinitions[dataproviderDefinitions$shortmodelname==model,]$dataprovider
  dps<-paste("ARRAY['",dataprovider,"']",sep="")
  return (dps)	
}


getParameterString<-function(param){
  pdef <- parameterDefinitions[parameterDefinitions$miopdb_par==param,]
  valueparametername <- as.character(pdef$valueparametername)
  parameterstring<-paste("ARRAY['",valueparametername,"']",sep="")
  return (parameterstring)	
}


convertValues <- function(values,param){
  pdef <- parameterDefinitions[parameterDefinitions$miopdb_par==param,]
  miopdb_unit<- as.character(pdef$miopdb_unit)
  unit<- as.character(pdef$unit)
  if (!is.na(unit)&&!miopdb_unit==unit)
    values <- ud.convert(values,unit,miopdb_unit)
  return(values)
}
  
getLevelString<-function(param,lev){
  levelstring <- "NULL"
  return(levelstring)
  pdef <- parameterDefinitions[parameterDefinitions$miopdb_par==param,]

  if (nrow(pdef)!=0) {
    if (is.null(lev)){
      levelparametername <- as.character(pdef$groundlevelparametername)
      level <- as.character(pdef$groundlevel)
    } else{
      levelparametername <- as.character(pdef$levelparametername)
      levpdef <- levelParameterDefinitions[levelParameterDefinitions$levelparametername==levelparametername,]
      #check if we need to convert level from miopdbunit
      if (nrow(levpdef)!=0) {
        miopdb_levelunit<- as.character(levpdef$miopdb_unit)
        levelunit<- as.character(levpdef$unit)
        if (!is.na(levelunit)&&!miopdb_levelunit==levelunit){
          lev <- ud.convert(lev,miopdb_levelunit,levelunit)
        }
      }
      level <- format(lev,scientific=FALSE)
      
    }

    
    levelstring <- paste("'",level," ",levelparametername,"'",sep="")  
  }
  return(levelstring)
}



convertLevels <- function(levs,param){
  pdef <- parameterDefinitions[parameterDefinitions$miopdb_par==param,]
  if (nrow(pdef)!=0) {
    levelparametername <- as.character(pdef$levelparametername)
    levpdef <- levelParameterDefinitions[levelParameterDefinitions$levelparametername==levelparametername,]
    #check if we need to convert level from miopdbunit
    if (nrow(levpdef)!=0) {
      miopdb_levelunit<- as.character(levpdef$miopdb_unit)
      levelunit<- as.character(levpdef$unit)
      if (!is.na(levelunit)&&!miopdb_levelunit==levelunit){
        levs <- ud.convert(levs,levelunit,miopdb_levelunit)
      }

    }
  }
  return(levs)
}
